Add fancy menu style to themes

stefonarch opened the pull request:
Issues:

* Valendas theme: opacity only gets displayed at second opening
* Silver, Ambiance: item text on hover gets border and background with widget style fusion mostly
* Didn't find out how to style QtoolButtons, this doesn't work:
```
LXQtFancyMenuWindow > QToolButton{
    background: #f0f0f0;
}
````
* Didn't add everywhere style for scrollbars, waiting for size to increase.

Removed also fixed tiny font size for worldclock in "light" theme.

gfgit commented:
I've adapted the KDE Plasma theme to FancyMenu, it's decent. I tried with the others but results are ugly at best :)
One thing I've noticed is that the structure of files is quite different between themes.
A part from the intrinsic differences of colors and styles, also item order and decription comments are different.

Also I've noticed Light theme (or maybe another) uses `palette(highlight)` instead of hadcoding color values. Could this be used to make common code for a bunch of themes and just set the correct palette at the beginninge?
Just an idea, I'm not familiar with QSS

stefonarch commented:
> I tried with the others but results are ugly at best :)

Hm, are you sure (except scrollbars)? They should be quite similar to the their mainmenu counterparts. Some themes (frost ,light, KDE) require the default bright palette.

![screen_area_mar_08:23:28_](https://github.com/lxqt/lxqt-themes/assets/10681413/443a5af8-0bab-4b79-a991-

![screen_area_mar_08:24:14_](https://github.com/lxqt/lxqt-themes/assets/10681413/b3119616-50e4-4dc9-99e6-291a605f1947)
34b1e142dcb3)



>One thing I've noticed is that the structure of files is quite different between themes.

Yes, they come from different authors and had some fixes too time ago.

>Also I've noticed Light theme (or maybe another) uses palette(highlight) instead of hardcoding color values. Could this be used to make common code for similar themes and just set the correct palette at the beginning?

Palette is not set in themes, some theme has its palette counterpart with the same name, we had the idea that there could be a check button "apply corresponding palette" in `lxqt-config-appearance` but it's not implemented yet. This would style at least the colors of qps, pcmanfm-qt, lximage and others.

stefonarch commented:
Noticed only now a bug in ambiance looking at the screenshot here.

gfgit commented:
![light_theme](https://github.com/lxqt/lxqt-themes/assets/42845724/b359bcb4-0b66-4b57-8348-9cba21791022)

With "Light" theme the "All Application" Category is slightly moved to right

gfgit commented:
Also why do I have correct FancyMenu panel icon on "System" theme even if you did not touch it's `lxqt-panel.qss` file?

stefonarch commented:
> With "Light" theme the "All Application" Category is slightly moved to right

This was  an indention on hover which I didn't understand where it comes from, but I fixed it, here it's not present anymore.

> applied the normal "System" theme which does not have an icon yet

Fixed, thanks.

> What I meant here is that some styles are quite similar. I think KDE Plasma, Arch Colors and Dark are quite the same thing but with different palette. So if we make them use placeholders like palette(window) instead of hard-coding colors we end up with 3 almost identical lxqt-panel.qss files.

This has a point, but if an user has a custom or different palette themes are completely different, and it will probably break appearance if introduced. But there is much which could  be improved, I agree.

tsujan commented:
> Valendas theme: opacity only gets displayed at second opening

Not here — unless I've misunderstood your point.

> Silver, Ambiance: item text on hover gets border and background with widget style fusion mostly

I don't see that here, but if you mean that the *focused* item gets a border and background with Fusion (which does here), add this
```css
#FancyMenu QListView:focus {
    outline: transparent;
}
```

Anyway, the code needs some changes for a better styling. I'll make a PR and explain things in it later.

stefonarch commented:
`#FancyMenu QListView:focus `works in Ambiance, but in change "Favorites" got an indentation, same as in "light" theme indeed mentioned by @gfgit .
Fixed all for the moment. I'll wait for the PR.

stefonarch commented:
#AppView and scrollbars are styled now, except for light themes - IMO the default fusion scrollbar doesn't look too bad there. @gfgit ?

Missing/Issues:

* added transparency in dark as in main menu, but same issue: it works only at second opening of the menu
* Couldn't style the separator here which should have some margin IMO in some themes:

https://github.com/lxqt/lxqt-themes/assets/10681413/64b00945-af7e-4e72-9727-785ae39ceea1

* QToolButtons

tsujan commented:
> added transparency in dark as in main menu

Long story short, it isn't possible in a consistent way. If I find the time, I'll explain the long version later.

gfgit commented:
@stefonarch Arch Colors has poor contrast here. My icon theme Papirus Light, tried also Papirus dark but it doesn't affect the theme colors.
Still have to test other themes

![arch_colors_lxqt_panel](https://github.com/lxqt/lxqt-themes/assets/42845724/23c15b37-71e0-4541-95ef-377af3ad9af1)

gfgit commented:
> Make sure to restart the panel module after widget changes (theme changing for and back works too), here it looks like that, except the white square):

Yes, but also I forgot to apply commits in `master` and `fancy_menu_scrollbar`.
Is your screenshot on Arch Colors? It looks good!
You have white scrollbar corner widget, the little square in bottom right corner. I don't have it, it's same color as the view background, don't know why this difference but I prefer it to be the same color.
Also the buttons with white square around them I think are a bit out of place, they should be darker

gfgit commented:
Corner widget is noticeable also on Clearlooks theme because scrollbars are light gray compared to white background.
Ambience, Frost, KDE Plasma, Kvantum, Leech, Valendas have same white background on buttons.
Dark and Light themes are really nice!

On Valendas the view frame is a bit weird having only vertical borders and not horizontal but if it's done on purpose then ok.

gfgit commented:
Sometimes on Dark theme I get a horizontal scrollbar on bottom of category view which is not actually needed because none of the items are cut out. Restarting panel again makes it disappear

stefonarch commented:
> Yes, but also I forgot to apply commits in `master` and `fancy_menu_scrollbar`.
> Is your screenshot on Arch Colors? It looks good!

O sure, that was the reason!
I'm the author of Arch-colors, Silver and Valendas, long time ago, thanks ;)

Actual screenshot is this (on https://github.com/lxqt/lxqt-panel/pull/1983 which has scrollbars only in app view. 

![screen_area_mer_12:31:59_](https://github.com/lxqt/lxqt-themes/assets/10681413/6ea94ac2-f507-47fd-9168-d4402b5893b6)

We could style the scrollbars in the light theme a little bit, at least handle color IMO.

stefonarch commented:
> Ambience, Frost, KDE Plasma, Kvantum, Leech, Valendas have same white background on buttons. 

This is the issue already mentioned, it needs a some change to style those QToolButton backgrounds.

> On Valendas the view frame is a bit weird having only vertical borders and not horizontal but if it's done on purpose then ok.

I've to review it as transparency doesn't work on first opening.

stefonarch commented:
How about that scrollbar style for light themes?


https://github.com/lxqt/lxqt-themes/assets/10681413/4f6c057e-f7fb-4636-a563-16c38ed4baff

gfgit commented:
> This is the issue already mentioned, it needs a some change to style those QToolButton backgrounds.

I did the following for KDE Plasma theme:

```
#FancyMenu QToolButton {
    margin: 3px;
    border-radius: 4px;
    border: 1px solid transparent;
    padding: 4px;
}

#FancyMenu QToolButton:hover {
    border: 1px solid rgba(30, 145, 255, 100%);
}

#FancyMenu QToolButton:pressed {
    color: rgba(54, 54, 54, 100%);
    border: 1px solid rgba(30, 145, 255, 100%);
    background-color: rgba(30, 145, 255, 30%);
}

#FancyMenu QToolButton::menu-indicator {
    image: none;
}
```

Just copied from Calendar and Volume plugins, not sure if last `menu-indicator` is really needed.
Also did not really understand the `color` variable but results are nice flat button with borders on hover and semi transparent fill on pressed.

stefonarch commented:
> To achieve this you press on first button then move mouse over second button, release and press again. First button stays "hovered/focused" while second is pressed.

This doesn't work in labwc (IMO more correct). `menu-indicator` isn't needed afaik.

stefonarch commented:
Ready for review now.

gfgit commented:
Tested again, Kvantum needs the `outline: transparent` fix like you did to Ambience

tsujan commented on themes/Arch-Colors/lxqt-panel.qss:
```diff
@@ -179,7 +179,128 @@ QMenu::indicator:non-exclusive:checked {
+    background: #003d5b;
+    color: #d7e3e6;
+    padding: 0px;
+}
```
IMO, shifting to right on hovering is a little disturbing. Is it intentional? If not, please change the above two blocks to these (also a 30-px right padding seems unnecessary, but you could keep it if you like it):

```xml
#FancyMenu QListView::item {
    color: #d7e3e6;
    padding: 5px 3px 5px 3px;
    border: 0px;
}

#FancyMenu QListView::item:hover {
    background: #003d5b;
    color: #d7e3e6;
}
```

tsujan replied:
> This padding: 0px was an illogical hack

Haha... CSS sometimes seems illogical, yes. However, it's a matter of inheritance: when something isn't specified, it might be inherited from the Qt style and cause confusion.

stefonarch replied:
btw I just  copied the themes from mainmenu and replaced `Main → Fancy` and ` Action → QList` and removed usused stuff.

tsujan replied:
Not all widgets get similar details from the Qt style, although their names may be the same in CSS — Qt's code is quite complicated. However, please check if the main menu has a similar problem.

In short, if you see an "unexplainable" effect with CSS, you could first search for another part of CSS that sets a value and then, if there isn't any, search for totally missing values.

stefonarch replied:
> if there isn't any, search for totally missing values

that was the point, but it's still illogical ;)

tsujan commented on themes/Arch-Colors/lxqt-panel.qss:
```diff
@@ -179,7 +179,128 @@ QMenu::indicator:non-exclusive:checked {
+    color: white;
+    background:  #041b27;
+    padding: 3px;
+}
```
By not setting the border here, it's left to the widget style. It may seem OK with Fusion but can be quite weird with Kvantum (and, perhaps, some other styles). Please add it like this (I chose it from the theme itself, but could change it if you want):

```css
#FancyMenu QLineEdit {
    color: white;
    background:  #041b27;
    padding: 3px;
    border: 1px solid #d7e3e6;
}
```

tsujan replied:
> border: none;

You prefer no border with the search entry? It's OK with me if you do -- just asked to know if it's intentional.

stefonarch replied:
There was a border in theme colors, but it was drawn by the widget. It's intentional.

stefonarch commented:
> Tested again, Kvantum needs the `outline: transparent` fix like you did to Ambience

Done, thanks!

tsujan commented on themes/Clearlooks/lxqt-panel.qss:
```diff
@@ -290,6 +298,151 @@ QMenu::indicator:non-exclusive:checked:selected {
+    border: 1px solid transparent;
+    padding: 0px;
+    margin: 0px;
+}
```
`color: #1a1a1a;` is missing here. Without it, there's light text on light background with dark Qt styles.

stefonarch replied:
Clearlooks is weird with dark palettes anyway (volume and brightness popup follow palette and are dark), so I jumped this, but will add it. IMO we should load the default theme when clearlooks, frost, light or kde is chosen. https://github.com/lxqt/lxqt-config/issues/974

tsujan replied:
IMO, the default palettes should to be avoided as far as possible -- unless in themes like the system theme, and if I correctly remember, in Frost, which is supposed to inverse them.

tsujan commented on themes/Valendas/lxqt-panel.qss:
```diff
@@ -176,6 +176,119 @@ QMenu QToolButton {
+    border-right: 1px solid white;
+    background-color: #444444;
+    margin: 1px;
+}
```
If you don't want top and bottom borders, you need to add this (otherwise, they might be taken from Qt styles like Kvantum and the result will be ugly):

```css
    border-top: 1px solid transparent;
    border-bottom: 1px solid transparent;
```

If not, then

```css
#FancyMenu #AppView {
    border: 1px solid white;
    background-color: #444444;
    margin: 1px;
}
```

stefonarch replied:
Done, in fact it wasn't working and I forgot about, I forgot that it needs all 4. 
White was ugly, they are greyish. Found another issue, it needs `outline:transparent` too.

And another big issue: with "windows" widget style all scrollbars are ugly in all themes.

tsujan commented on themes/ambiance/lxqt-panel.qss:
```diff
@@ -311,6 +311,134 @@ QCalendarWidget QWidget {
+#FancyMenu QListView::item {
+    color: #f2f1f0;
+    padding: 5px 0px 5px 14px;
+}
```
Shift to right is still visible with some Kvantum style themes -- it's only by chance that it may not be visible with Fusion because padding is changed below.

Now, there are two options. If you want to keep `padding: 5px 0px 5px 14px;` (which looks OK), make it as

```css
#FancyMenu QListView::item {
    color: #f2f1f0;
    padding: 5px 0px 5px 14px;
    border: 0px;
}
```

and *remove `padding: 0px;` from the following, related blocks*.

But if you don't wan the padding, just do as

```css
#FancyMenu QListView::item {
    color: #f2f1f0;
    padding: 0px;
    border: 0px;
}
```

stefonarch replied:
Done with padding.

tsujan commented on themes/dark/lxqt-panel.qss:
```diff
@@ -458,6 +457,123 @@ QMenu::indicator:non-exclusive:checked {
+    margin-bottom: 1px;
+    color: white;
+    selection-background-color: #640b0c;
+}
```
By not adding border, you've left it to the widget style. It may be OK with Fusion, but it isn't OK with some Kvantum themes (which are light and have completely rounded borders). Please change it as

```css
#FancyMenu QLineEdit {
    background: black;
    border: 1px solid #3c3c3c;
    padding: 3px;
    margin-bottom: 1px;
    color: white;
    selection-background-color: #640b0c;
}
```

I've just added `border: 1px solid #3c3c3c;` from `#MainMenu QLineEdit`.

tsujan commented on themes/silver/lxqt-panel.qss:
```diff
@@ -274,6 +274,148 @@ QCalendarWidget #qt_calendar_prevmonth {
+
+#FancyMenu QListView::item:focus {
+    padding: 0px;
+}
```
This causes a shift because the padding was 2px above. Please remove it.

tsujan commented:
OK, my review is finished. Will review it again after you made changes.

stefonarch commented on themes/Leech/lxqt-panel.qss:
```diff
@@ -269,6 +269,152 @@ QMenu::left-arrow:selected {
+    color: #ffffff;
+    margin-left: 1px;
+    padding-left: 7px;
+}
```
As I said: this is all just copy&paste from mainmenu, but in fact here it _does_ shift while in mainmenu not, I didn't spot it. I'll remove it in mainmenu too. 

You missed a closing ;  at ` padding: 0px`

tsujan replied:
> You missed a closing ; at  padding: 0px

Yes :D

stefonarch commented:
With "windows" widget style  
```
#FancyMenu QListView::focus {
    outline:transparent;
    padding: 0px;
}
``` 
has to be added where it's not present already, like dark and maybe others. The scrollbars are ugly in all dark themes while in the light ones it's ok. I tried to set a background color instead of `transparent`, no effect.

![screen_area_gio_21:36:08_](https://github.com/lxqt/lxqt-themes/assets/10681413/c0170c74-6812-40f9-b660-0812c2ccb19d)

tsujan commented:
Which theme do you mean?

tsujan commented:
I've got CSS sickness tonight. I'll review it again tomorrow.

stefonarch commented:
> Which theme do you mean?

screenshot is dark theme here. But scrollbars are like that in all dark themes. Got CSS poisoned too.

tsujan commented:
Just a show-off without any LXQt theme and with Kvantum's translucency and transient scrollbars ;)

![kvantum](https://github.com/lxqt/lxqt-themes/assets/981076/f2ccbbc9-eb75-42e1-94c0-a3b92522a442)

tsujan reviewed (approved):
All issues I found are fixed. GTM.

IMHO, it's better to merge this and do other works in another PR. For example, we can set the view icon size for some themes to show users that they can be changed, like this:

```css
#FancyMenu QListView {
    icon-size: 24px;
}
```

stefonarch commented:
> All issues I found are fixed. GTM.
> 
> IMHO, it's better to merge this and do other works in another PR. For example, we can set the view icon size for some themes to show users that they can be changed, like this:
> 
> ```css
> #FancyMenu QListView {
>     icon-size: 24px;
> }
> ```
 I doubt users fiddle with those files, but we could add an option as there is also one for font size already.

AzumaHazuki commented:
So we can assume this deprecates `#WingMenuPlugin` going forward? I'm still making themes and want to know if I should keep styling for the Wing Menu or leave those sections out.

Also, have the refactored Leech and Clearlooks themes I uploaded been merged?

stefonarch commented:
Wingmenu style is useful only for people who are able to compile it themselves, as it can be never distributed legally.

> Also, have the refactored Leech and Clearlooks themes I uploaded been merged? 

No, there are so many changes there that nobody has reviewed it yet. You could explain maybe the differences and reasons, keep in mind that if you change aspects you will change themes people are using.

AzumaHazuki commented:
The changes are just to make them slightly more faithful to their original GTK2 counterparts, and to clean up the directory structure. These were noted in the PRs themselves, though it sounds like now I should be refactoring every theme I've ever made to have FancyMenu support?

Is the FancyMenu plugin even complete? It could use some more options, like placing the search field above or below, showing the user's avatar, and making it so mouse tracking works as it does on the Main Menu and Wing Menu instead of having to click a category to get it to change. The icon sizes for the categories, apps, and buttons below should also be customizable outside QSS. I really, really like where FancyMenu is going as I love WhiskerMenu, it's just not quiiiite there yet.

tsujan commented:
> Is the FancyMenu plugin even complete?

It's available to git users and, like any program, it'll evolve and be developed. For now, were're testing it and fixing any issues we encounter.

> It could use some more options....

Yes, of course. In the end, devs will decide which options are acceptable and when to stop adding options. It isn't supposed to be a clone of WhiskerMenu BTW.

AzumaHazuki commented:
I _think_ I've found all or almost all of the themeable elements of the widget as it is now: https://github.com/lxqt/lxqt/discussions/2517

Some things don't feel quite right, like addressing the main menu backdrop with `#FancyMenu > QWidget`, but overall this looks and feels very nice. I've been looking through the C++ code trying to understand how to move/change some things, but it's just too far above me.
